Infer imports for linking

slavapestov opened the pull request:
Completes the fix for <rdar://problem/39338239>.

slavapestov committed:
Sema: Infer '@_usableFromInline' imports by looking at references from inlinable functions

Completes the fix for <rdar://problem/39338239>.

slavapestov committed:
AST: Improve silly doc comment for ModuleDecl::getImportedModulesForLinking()

slavapestov commented:
@swift-ci Please test

slavapestov commented:
The 32-bit failure in the macOS full test is not caused by this change. I have a fix for it in https://github.com/apple/swift/pull/16302.

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test macOS

jrose-apple commented on include/swift/AST/Module.h:
```diff
@@ -373,8 +373,9 @@ class ModuleDecl : public DeclContext, public TypeDecl {
-  /// Extension of the above hack. Identical to getImportedModulesForLookup()
-  /// for imported modules, otherwise also includes @usableFromInline imports.
+  /// Extension of the above hack. Behaves identically to
+  /// getImportedModulesForLookup() for Clang modules, otherwise also
```
I still think you shouldn't mention the Clang part at all. Why does someone care about that when using the API?

jrose-apple commented on include/swift/AST/Module.h:
```diff
@@ -783,7 +784,7 @@ class SourceFile final : public FileUnit {
   ///
   /// This is filled in by the Name Binding phase.
-  ArrayRef<std::pair<ModuleDecl::ImportedModule, ImportOptions>> Imports;
+  MutableArrayRef<std::pair<ModuleDecl::ImportedModule, ImportOptions>> Imports;
```
Can you tweak the comment to say why it's mutable?

jrose-apple commented on lib/Sema/TypeCheckAvailability.cpp:
```diff
@@ -2509,12 +2509,19 @@ bool AvailabilityWalker::diagAvailability(const ValueDecl *D, SourceRange R,
+      auto *SF = cast<SourceFile>(DC->getModuleScopeContext());
+      auto *M = D->getDeclContext()->getParentModule();
+      if (SF->getParentModule() != M)
+        SF->markUsableFromInlineImport(M);
```
The owning module might not be imported directly (for example, AppKit re-exports Foundation, and `_exported` imports elsewhere in the module are also visible). You'll need something a little more powerful here—it might just come down to adding additional imports.

slavapestov replied:
Oh, adding an additional import should be OK as long as we only do it once.

jrose-apple replied:
I'm not sure what you mean by "only do it once". It'll happen once per indirect import.

jrose-apple reviewed (commented):
This still has a hole in it, sorry.

slavapestov committed:
Fix '@_usableFromInline import' inference interaction with '@_exported import'

A module might be visible indirectly, because it's @_exported by
one of the modules we import. In this case, we have to add it to our
direct list of imports so that it can be autolinked.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@jrose-apple Looks like adding an import like that breaks name lookup. Perhaps UsableFromInline imports should not be considered by name lookup at all? Or maybe instead of adding a new import, we find the one through which we transitively found the referenced declaration. I’ll let you figure something out.

jrose-apple commented:
Ah, right, modules can shadow declarations that would otherwise be re-exported. Hm.
